taichi/backends/opengl/opengl_api.cpp

```diff
@@ -474,24 +496,24 @@ struct CompiledProgram::Impl {
         cpit++;
       }  // extract back to all extptr from my baseptr
     }
+    if (used.print) {
+      auto rt_buf = (GLSLRuntime *)launcher->impl->runtime_ssbo->map();
+      auto msg_count = rt_buf->msg_count;
+      if (msg_count > MAX_PRINT_ENTRIES) {
+        TI_WARN(
+            "[glsl] Too much print within one kernel: {} > {}, cutting tail",
+            msg_count, MAX_PRINT_ENTRIES);
+        msg_count = MAX_PRINT_ENTRIES;
+      }
+      for (int i = 0; i < msg_count; i++) {
+        std::cout << rt_buf->msg_buf[i] << std::endl;
```

Comment:
How to distinguish `float` from `int`?

Reply:
1. An extra bitmap. Can be overhead-saving but hard to maintain, and can't extend to `f64` or `i64`.
2. An extra int array. Can cause large overhead copying message, but easy to extend, even to `string`.
3. An extra char array. No, OpenGL doesn't support `u8` at all.
WDYT?

Reply:
I like 2, since you mentioned `string` support as well. We can have a 32-bit integer for every data element to be printed. This 32-bit integer can encode both the type and the size.

For example, if you need to support `float`, `int` and `str`, then the type info needs 2bits, with the rest 30bits being the size

* For an integer `42`, type=`0`, size doesn't matter, since we know it's int32
* For a float `233.0`, type=`1`, size doesn't matter for the same reason.
* Flor a str `abcde`, type=`2`, size=`5`

Reply:
> 1MB

Does 1MB data really printable on a terminal? I think 32KB is enough.

Reply:
I don't expect the buffer to be fully filled for most of the time, but this is really a number I came up randomly. Please feel free to choose a reasonably large size (512bytes seems too small to me)